cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -0,0 +1,509 @@
+                            AdapterSpecification adapterSpec) {
+    ProgramType programType = adapterTypeInfo.getProgramType();
+    if (programType.equals(ProgramType.WORKFLOW)) {
+      Map<String, WorkflowSpecification> workflowSpecs = spec.getWorkflows();
+      for (Map.Entry<String, WorkflowSpecification> entry : workflowSpecs.entrySet()) {
+        Id.Program programId = Id.Program.from(namespaceId, adapterSpec.getType(), entry.getValue().getName());
+        deleteSchedule(programId, SchedulableProgramType.WORKFLOW, adapterSpec.getScheduleName());
+      }
+    } else {
+      // Only Workflows are supported to be unscheduled in the current implementation
+      throw new UnsupportedOperationException(String.format("Unsupported program type %s for adapter",
+                                                            programType.toString()));
+    }
+  }
+
+  // Adds a schedule to the scheduler as well as to the appspec
+  private void addSchedule(Id.Program programId, AdapterSpecification adapterSpec) {
+    String cronExpr = toCronExpr(adapterSpec.getProperties().get("frequency"));
+    Preconditions.checkNotNull(cronExpr, "Frequency of running the adapter is missing. Cannot schedule program");
+    Schedule schedule = new Schedule(adapterSpec.getScheduleName(), adapterSpec.getScheduleDescription(), cronExpr);
+    ScheduleSpecification scheduleSpec = new ScheduleSpecification(schedule,
+                                           new ScheduleProgramInfo(SchedulableProgramType.WORKFLOW, programId.getId()),
+                                           adapterSpec.getProperties());
+
+    scheduler.schedule(programId, scheduleSpec.getProgram().getProgramType(), scheduleSpec.getSchedule());
+    //TODO: Scheduler API should also manage the MDS.
+    store.addSchedule(programId, scheduleSpec);
+  }
+
+  // Deletes schedule from the scheduler as well as from the app spec
+  private void deleteSchedule(Id.Program programId, SchedulableProgramType programType, String scheduleName) {
+    scheduler.deleteSchedule(programId, programType, scheduleName);
+    //TODO: Scheduler API should also manage the MDS.
+    store.deleteSchedule(programId, programType, scheduleName);
+  }
+
+  // Sources for all adapters should exists before creating the adapters.
+  private void validateSources(String adapterName, Set<Source> sources) throws IllegalArgumentException {
+    // Ensure all sources exist
+    for (Source source : sources) {
+      if (Source.Type.STREAM.equals(source.getType())) {
+        if (!streamExists(source.getName())) {
+          throw new IllegalArgumentException(String.format("Stream %s must exist during create of adapter: %s",
+                                                           source.getName(), adapterName));
+        }
+      } else {
+        throw new IllegalArgumentException(String.format("Unknown Source type: %s", source.getType()));
```

Comment:
Usually I prefer failure condition closer to the condition statement to make it clear what fails and also have less nested code structure. E.g.

```java
if (!Source.Type.STREAM.equals(source.getType())) {
  throw new IllegalArgumentException(...);
}
if (!streamExists(source.getName()) {
  throw new IllegalArgumentException(...);
}
```

In fact, for the check in here, you can turn it into two `Preconditions` check:

```java
Preconditions.checkArgument(Source.Type.STREAM.equals(source.getType()), "Stream %s must ...", source.getName());
Preconditions.checkArgument(...);
```